### PR TITLE
implement size_hint on sys iterators

### PR DESCRIPTION
**Pull request by Dylan-DPC-zz**
Addresses #49205

Rework of #49552

### CODE REVIEW

**Comment by Dylan-DPC** [comment]
Marking as draft because i'm expecting some CI failures

File: src/libstd/sys/cloudabi/shims/fs.rs [hunk 1]
```diff
@@ -140,6 +140,10 @@ impl Iterator for ReadDir {
     fn next(&mut self) -> Option<io::Result<DirEntry>> {
         match self.0 {}
     }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(0))
```
Comment (sfackler) [inline src/libstd/sys/cloudabi/shims/fs.rs:145]: This type is uninhabited - does it really need a size_hint override?

Comment (Dylan-DPC) [reply src/libstd/sys/cloudabi/shims/fs.rs:145]: Yeah that makes sense to remove it. I just ported the old pr :D

File: src/libstd/sys/cloudabi/shims/net.rs [hunk 2]
```diff
@@ -299,6 +299,9 @@ impl Iterator for LookupHost {
     fn next(&mut self) -> Option<SocketAddr> {
         match self.0 {}
     }
+    fn size_hint(&self) -> (usize, Option<usize>) {
```
Comment (sfackler) [inline src/libstd/sys/cloudabi/shims/net.rs:302]: All of these are uninhabited as well.

**Comment by Dylan-DPC** [comment]
I don't see much gain from this though.. guess i'll close it for now :D

**Pull request closed without merging**